### PR TITLE
feat: I/O safety for 'sys/sendfile'

### PR DESCRIPTION
**Pull request by SteveLauC**
#### What this PR does:
1. Adds I/O safety for module `sys/sendfile`.

### CODE REVIEW

File: src/sys/sendfile.rs [hunk 1]
```diff
@@ -23,16 +23,23 @@ use crate::Result;
 /// For more information, see [the sendfile(2) man page.](https://man7.org/linux/man-pages/man2/sendfile.2.html)
 #[cfg(any(target_os = "android", target_os = "linux"))]
 #[cfg_attr(docsrs, doc(cfg(all())))]
-pub fn sendfile(
-    out_fd: RawFd,
-    in_fd: RawFd,
+pub fn sendfile<F1: AsFd, F2: AsFd>(
```
Comment (SteveLauC) [inline src/sys/sendfile.rs:26]: Two generic types are needed as they may be different types when used.

File: test/test_sendfile.rs [hunk 2]
```diff
@@ -23,7 +24,12 @@ fn test_sendfile_linux() {
 
     let (rd, wr) = pipe().unwrap();
     let mut offset: off_t = 5;
-    let res = sendfile(wr, tmp.as_raw_fd(), Some(&mut offset), 2).unwrap();
+    // The construct of this `OwnedFd` is a temporary workaround, when `pipe(2)`
+    // becomes I/O-safe:
+    // pub fn pipe() -> std::result::Result<(OwnedFd, OwnedFd), Error>
+    // then it is no longer needed.
+    let wr = unsafe { OwnedFd::from_raw_fd(wr) };
+    let res = sendfile(&wr, &tmp, Some(&mut offset), 2).unwrap();
```
Comment (SteveLauC) [inline test/test_sendfile.rs:32]: This is a temporary workaround to cope with the existing `pipe(2)` interface and should be resolved in other PRs.

**Comment by SteveLauC** [comment]
The CI failure is not code-related, it is a network issue:

```
warning: spurious network error (1 tries remaining): SecureTransport error: (null); class=Net (12)
error: failed to get `bitflags` as a dependency of package `nix v0.26.1 (/private/var/folders/v7/fs2b0v3s0lz1n57gj9y4xb5m0000gn/T/cirrus-ci-build)`

Caused by:
  failed to load source for dependency `bitflags`

Caused by:
  Unable to update registry `crates-io`

Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`

Caused by:
  network failure seems to have happened
  if a proxy or similar is necessary `net.git-fetch-with-cli` may help here
  https://doc.rust-lang.org/cargo/reference/config.html#netgit-fetch-with-cli

Caused by:
  SecureTransport error: (null); class=Net (12)

Exit status: 101
```

File: src/sys/sendfile.rs [hunk 3]
```diff
@@ -23,16 +23,23 @@ use crate::Result;
...
+    out_fd: &F1,
```
Comment (asomers) [inline src/sys/sendfile.rs:27]: As discussed elsewhere, should we make these `F1` instead of `F1`?

**Review by asomers — APPROVED** [review]
bors r+

**Pull request merged**